src/safe_op.hpp

```diff
@@ -331,4 +332,103 @@ namespace Safe
         return num < 0 ? -num : num;
     }
 
+    namespace Internal
+    {
+        template <typename from, typename to, typename = void>
+        struct is_safely_convertible : std::false_type
+        {
+            static_assert(std::is_integral<from>::value&& std::is_integral<to>::value,
+                          "from and to must both be integral types");
+        };
+
+        template <typename from, typename to>
+        struct is_safely_convertible<
+            from, to,
+            typename std::enable_if<((std::numeric_limits<from>::max() <= std::numeric_limits<to>::max()) &&
+                                     (std::numeric_limits<from>::min() >= std::numeric_limits<to>::min()))>::type>
+            : std::true_type
+        {
+        };
+
+        template <typename T, typename U, typename = void>
+        struct have_same_signedness : std::false_type
+        {
+            static_assert(std::is_integral<T>::value&& std::is_integral<U>::value,
+                          "T and U must both be integral types");
+        };
+
+        // SFINAE overload for (T signed and U signed) or (T unsigned and U unsigned)
+        // note: !a != !b == a XOR b (for a, b bool)
+        template <typename T, typename U>
+        struct have_same_signedness<
+            T, U, typename std::enable_if<!((!std::is_signed<T>::value) != (!std::is_signed<U>::value))>::type>
```

Comment:
This works fine:

```
std::enable_if<(std::is_signed<T>::value == std::is_signed<U>::value)>::type>
```

Reply:
too obvious after being awake for 20hrs... Thanks!